libraries/AP_HAL_Linux/RCOutput_Bebop.cpp

```diff
@@ -0,0 +1,440 @@
+    ret = pthread_cond_init(&_cond, NULL);
+
+    if(ret != 0) {
+        perror("RCout_Bebop: failed to init cond\n");
+        goto err_mutex;
+    }
+
+    ret = pthread_attr_init(&attr);
+    if(ret != 0) {
+        perror("RCOut_Bebop: failed to init attr\n");
+        goto err_mutex;
+    }
+    pthread_attr_setinheritsched(&attr, PTHREAD_EXPLICIT_SCHED);
+    pthread_attr_setschedpolicy(&attr, SCHED_FIFO);
+    pthread_attr_setschedparam(&attr, &param);
+    ret = pthread_create(&_thread, &attr, _control_thread, this);
+    if(ret != 0) {
+        perror("RCOut_Bebop: failed to create thread\n");
+        goto err_mutex;
+    }
+
+    _clear_error();
+
+    /* Set an initial dummy frequency */
+    _frequency = 50;
+    return;
+
+err_mutex:
+    pthread_mutex_unlock(&_mutex);
+    return;
+}
+
+void LinuxRCOutput_Bebop::set_freq(uint32_t chmask, uint16_t freq_hz)
+{
+    _frequency = freq_hz;
+}
+
+uint16_t LinuxRCOutput_Bebop::get_freq(uint8_t ch)
+{
+    return _frequency;
+}
+
+void LinuxRCOutput_Bebop::enable_ch(uint8_t ch)
+{
+}
+
+void LinuxRCOutput_Bebop::disable_ch(uint8_t ch)
+{
+    _stop_prop();
+}
+
+void LinuxRCOutput_Bebop::write(uint8_t ch, uint16_t period_us)
+{
+
+    pthread_mutex_lock(&_mutex);
+    _period_us[ch] = period_us;
+    pthread_cond_signal(&_cond);
+    pthread_mutex_unlock(&_mutex);
+}
+
+void LinuxRCOutput_Bebop::write(uint8_t ch, uint16_t* period_us, uint8_t len)
+{
+    for (int i = 0; i < len; i++)
+        write(ch + i, period_us[i]);
+}
+
+uint16_t LinuxRCOutput_Bebop::read(uint8_t ch)
+{
+    if(ch < BEBOP_BLDC_MOTORS_NUM) {
+        return _period_us[ch];
+    }
+    else {
+        return 0;
+    }
+}
+
+void LinuxRCOutput_Bebop::read(uint16_t* period_us, uint8_t len)
+{
+    for (int i = 0; i < len; i++)
+        period_us[i] = read(0 + i);
+}
+
+void LinuxRCOutput_Bebop::set_esc_scaling(uint16_t min_pwm, uint16_t max_pwm)
+{
+    _min_pwm = min_pwm;
+    _max_pwm = max_pwm;
+}
+
+/* Separate thread to handle the Bebop motors controller */
+void* LinuxRCOutput_Bebop::_control_thread(void *arg) {
+    LinuxRCOutput_Bebop* rcout = (LinuxRCOutput_Bebop *) arg;
+
+    rcout->_run_rcout();
+    return NULL;
+}
+
+void LinuxRCOutput_Bebop::_run_rcout()
+{
+    uint16_t current_period_us[BEBOP_BLDC_MOTORS_NUM];
+    uint8_t i;
+    int ret;
+    struct timespec ts;
+
+    memset(current_period_us, 0, sizeof(current_period_us));
+
+    while(true) {
+        ret = clock_gettime(CLOCK_REALTIME, &ts);
+        if(ret != 0)
+            printf("failed to get time\n");
+
+        if(ts.tv_nsec > (1000000000 - BEBOP_BLDC_TIMEOUT_NS))
+        {
+            ts.tv_sec += 1;
+            ts.tv_nsec = ts.tv_nsec + BEBOP_BLDC_TIMEOUT_NS - 1000000000;
+        }
+        else {
+            ts.tv_nsec += BEBOP_BLDC_TIMEOUT_NS;
+        }
+
+        ret = 0;
+        while((memcmp(_period_us, current_period_us, sizeof(_period_us)) == 0) && (ret == 0))
+            ret = pthread_cond_timedwait(&_cond, &_mutex, &ts);
```

Comment:
> The mutex is taken when we enter the condition wait, either by the init function, or by the call to pthread_mutex_lock at the end of the loop. I know this makes the code a bit unclear but it works.

It still looks wrong to me. The `init()` function runs in one thread. This function runs in another thread with the following call chain:
```
LinuxRCOutput_Bebop::init()
    pthread_create()  // <------------ from now on, code below is in another thread
        LinuxRCOutput_Bebop::_control_thread()
            LinuxRCOutput_Bebop::_run_rcout()
```
You can't call `pthread_cond_timedwait()` without holding the mutex in the particular thread you are calling it. From PTHREAD_COND_TIMEDWAIT(3P):

```
       The pthread_cond_timedwait() and  pthread_cond_wait()  functions  shall
       block  on a condition variable. The application shall ensure that these
       functions are called with mutex locked by the  calling  thread;  other‐
       wise,  an  error  (for  PTHREAD_MUTEX_ERRORCHECK and robust mutexes) or
       undefined behavior (for other mutexes) results.
```

 And you can't call `memcmp()` without holding the mutex neither. So you should actually move the `pthread_mutex_lock()` call from the end of your loop to before this while loop.  The only undefined behavior here occurs on the first iteration of the loop, since for the second and so on you are taking the lock at the end. I bet you can just move the pthread_mutex_lock() call and everything will continue to work, but without the undefined behavior in the first iteration.

> By the way, Tridge pointed the fact that values are read/written atomically, and so the mutex is probably not needed.

Without the mutex how would you wake up the other thread? We certainly don't want the other thread to be busy looping. And, if it was a single `uint16_t` then it would be ok. However you have an array here and reading and writing to it is not atomic. The race window is very small, but it's there.